Update users report

rikeoka opened the pull request:
* Update user access reports for compliance
* Add new_users and users to CsvExportController
* Refactor User.to_csv to support requested options
* User report now work on deleted users, report effective project permissions, report on specific project's users or specific user's projects
* Admins can access the reports the same way under /admins/users#index 
* New links for admins are available to report on user's projects at /admin/users#show or project's users at /project/users#index
* Non-admins but authenticated users (compliance, etc) have the ability to generate the report at /csv_exports#new_users.  
* Relaxed permissions restriction to any logged in user for user reports since no sensitive data is present in report
* Updated tests

![screen shot 2016-07-12 at 11 56 51 am](https://cloud.githubusercontent.com/assets/11447948/16783634/b2519cbe-4839-11e6-9f5f-ce3ba9a4895e.png)
Options at csv_exports#new_users

/cc @zendesk/samson

### Tasks
 - [ ] :+1: from team

### References
 - Jira link: https://zendesk.atlassian.net/browse/SAMSON-261

### Risks
- Level: Low

grosser commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -19,6 +19,19 @@ def new
+    options = user_filter
+    respond_to do |format|
+      format.csv do
+        options[:datetime] = Time.now.strftime "%Y%m%d_%H%M"
```
move this into user_filter method ?

grosser commented on app/controllers/csv_exports_controller.rb:
```diff
@@ -52,6 +65,15 @@ def download
+  def user_filter
+    options = {}
+    options[:inherited] = params[:inherited] == "true"
+    options[:deleted] = params[:deleted] == "true"
```
prefer a single hash

```
{
  deleted: xyz,
  foobar: xyz
```

grosser replied:
ah nvm ... project and user should not be set ...

grosser commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+    inherited: false, deleted: false, project_id: nil, user_id: nil,
+    datetime: (Time.now.strftime "%Y%m%d_%H%M")
+  )
+    inherited = true if project_id || user_id
```
feels strange that it overrides that option ...

rikeoka replied:
Had this since when reporting on a specific project it would be a bit off to see anywhere between less than 100% of the users.  
Same with reporting on a specific user to see less than 100% of the projects, since most of their privileges are associated with their system permissions.

grosser commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+    datetime: (Time.now.strftime "%Y%m%d_%H%M")
+  )
+    inherited = true if project_id || user_id
+    users = if user_id.nil?
```
looks like it's finding deleted users when giving id and not deleted ... how about:

```Ruby
users = (deleted ? User.unscoped : User)
users = users.order(:id)
users = users.where(id: user_id) if user_id
```

grosser commented on app/models/user.rb:
```diff
@@ -136,4 +176,12 @@ def object_attrs_for_paper_trail(attributes)
     self.token = SecureRandom.hex
   end
+
+  private_class_method def self.permissions_count(project_id, inherited, users, permissions_projects = [])
```
don't need a default value for permissions_projects

grosser replied:
this look copy-pasty ... seems to duplicate logic from above ... can we reuse that somehow ?

grosser commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+      user_id && User.unscoped.where(id: user_id)
+    end
+    if inherited
+      permissions_projects = project_id ? Project.where(id: project_id) : Project.all
```
no need for `.all` ?

grosser commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+    end
+  end
+
+  def csv_line(project = nil)
```
no need for default

grosser commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+    if project
+      [id, name, email, project.id, project.name, effective_project_role(project), deleted_at]
+    else
+      [id, name, email, "", "SYSTEM", role.name, deleted_at]
```
I'd prefer something like this to avoid the duplication

```Ruby
[
  id,
  name,
  (project ? project.id : "")
  ...
]
```

grosser commented on app/models/user.rb:
```diff
@@ -121,6 +152,15 @@ def project_role_for(project)
   end
 
+  def effective_project_role(project)
+    project_role = project_role_for(project)
```
project_role_for can move into the else

grosser commented on app/models/user.rb:
```diff
@@ -121,6 +152,15 @@ def project_role_for(project)
+    if role_id == Role::SUPER_ADMIN.id
+      Role::ADMIN.name
+    else
+      role_id >= (project_role.try(:role_id) || 0) ? role.name : project_role.role.name
```
this logic already exists somewhere ... please reuse :)

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+      <div class="form-group">
+        <%= label_tag :deleted, "Include deleted users", class: "col-lg-2 control-label" %>
+        <div class="col-lg-2">
+          <%= select_tag :deleted, options_for_select([['No', false], ['Yes', true]], false), class: "form-control" %>
```
checkbox instead of a select ?

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+      <div class="form-group">
+        <%= label_tag :inherited, "Show inherited project permissions", class: "col-lg-2 control-label" %>
+        <div class="col-lg-2">
+          <%= select_tag :inherited, options_for_select([['No', false], ['Yes', true]], false), class: "form-control" %>
```
checkbox instead of select ?

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+      </div>
+      <div class="form-group">
+        <%= label_tag :user_id, "User", class: "col-lg-2 control-label" %>
+        <% users = [['', nil]] + User.pluck(:name, :id) +
```
should we sort by name ?

grosser commented on config/routes.rb:
```diff
@@ -86,7 +86,12 @@
   get '/logout', to: 'sessions#destroy'
 
-  resources :csv_exports, only: [:index, :new, :create, :show]
+  resources :csv_exports, only: [:index, :new, :create, :show] do
```
this looks unsymmetric ... can we re-use new and index with some kind of parameter ... otherwise create a new controller and make the existing one

grosser commented on test/models/user_test.rb:
```diff
@@ -187,9 +187,56 @@
+    end
+  end
+
+  describe "#csv_line and #effective_project_role" do
```
split these into 1 block per method ?

grosser commented on test/models/user_test.rb:
```diff
@@ -187,9 +187,56 @@
-  describe ".to_csv" do
-    it "generates csv" do
-      User.to_csv.split("\n").size.must_equal User.count + 1 + UserProjectRole.count
+  describe ".to_csv and .permissions_count" do
```
permissions_count is private so no need to mention it here

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+<h1>Request a CSV Report</h1>
```
`Request a User CSV Report` ?

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+<section>
+  <div class="csv-users-filters clearfix">
+    <%= form_tag users_csv_exports_path(format: :csv), method: :get, class: "form-horizontal" do %>
+      <div><strong>User Permission Reports</strong></div>
```
this should be in the h1

sandlerr commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
+  )
+    inherited = true if project_id || user_id
+    users = if user_id.nil?
+      deleted ? User.unscoped.order(:id) : User.order(:id)
```
`unscoped` can do surprising things sometimes, I think it reflects your intention better to do `User.with_deleted { User.order(:id) }`

rikeoka replied:
I thought with_deleted just passed unscoped onto the Model so this would be simplier?  I might be wrong through https://github.com/grosser/soft_deletion/blob/master/lib/soft_deletion/core.rb#L20

sandlerr replied:
didn't realise that, no reason to change it then

sandlerr commented on app/models/user.rb:
```diff
@@ -54,22 +54,53 @@ def administrated_projects
-            user_project_role.project.name, user_project_role.role.name
-          ]
+      csv << ["id", "name", "email", "projectiD", "project", "role", "deleted at"]
+      users.find_each do |user|
```
did you have problems doing this in a single sql query `users.each` rather than `find_each`?

rikeoka replied:
find_each saves a few percent of the total queries, but I guess it doesn't matter too much.

sandlerr commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+      </div>
+      <div class="form-group">
+        <%= label_tag :project_id, "Project", class: "col-lg-2 control-label" %>
+        <% projects = [['', nil]] + Project.pluck(:name, :id) %>
```
I kinda prefer using `permalink` over `id` here, I don't think we use project IDs much anywhere else

rikeoka replied:
Kept it id to be consistent with the deploys report, since that reports the permalink.  

If we really should be using permalink instead, I can update both reports in a separate PR.

sandlerr commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,46 @@
+      <div class="form-group">
+        <%= label_tag :user_id, "User", class: "col-lg-2 control-label" %>
+        <% users = [['', nil]] + User.pluck(:name, :id) +
+            User.unscoped.where("deleted_at IS NOT NULL").pluck(:name, :id).map { |u| ["(deleted) #{u[0]}", u[1]] } %>
```
does this need to be unscoped? Probably another good place to use `User.with_deleted { User.pluck(:name, :id) }`

sandlerr commented:
I think it would be awesome if it was clear to users which buttons (if any) will immediately generate a csv file and which ones will just link you to a menu page where you customise your export. Maybe using a slightly different style, symbol or text in those buttons.

rikeoka commented:
The download links on the various user pages are immediate with predefined params based on the page you're on.

Create buttons starts a job
Download buttons immediately generates and sends

Create links go to menu options:)

rikeoka commented:
@grosser updated.

@sandlerr user reports are generated real time.  From the respective user pages it'll generate the default reports for that page.  
From admin/users it'll generate report for all users no implied project permissions
From admin/users/:id, it'll generate report of a user's system role + implied project permissions (basically what's seen in the UI)
From project/:permalink/users, it'll generate report of all users for the project with implied project permissions (basically what's seen in the UI)
From csv_exports/new.html?type=users it'll ask users for report generation parameters and generate the report realtime.

sandlerr commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -76,16 +76,87 @@
-        p.save!(validate: false)
+      describe "as html" do
+        before do
+          p = projects(:test).dup
```
this should work without `dup`

rikeoka replied:
This was to create a second project to see if it appears in the dropdown.  I had a previous where it only showed one project in the UI.  It may be overkill to test UI logic.

sandlerr replied:
Cool, would you mind adding a comment to explain that? This is a slightly unconventional way to mock ActiveRecord objects

sandlerr commented on test/controllers/csv_exports_controller_test.rb:
```diff
@@ -76,16 +76,87 @@
+            csv_test({inherited: "true", deleted: "true"}, expected)
+          end
+
+          def csv_test(options = {}, expected = {})
```
could you set the default value for `expected` to `expected` rather than empty?

rikeoka replied:
since it only is called in the tests and tests should provide expected values I just removed the options.

sandlerr commented:
:+1:

rikeoka commented:
@grosser / @sandlerr 
Updated the algorithm for the inherited option for the users report.  By manually caching UserProjectRole for each user it reduces the number of single result queries from O(N*M) to O(N).

It requires the User.to_csv, User.csv_line and User.effective_project_role to have some logic that would otherwise natively be handled by Rails to be manually done, but I think in this case since it has a major performance improvement when generating a full report with a large number of projects to justify this approach.

Inclusive
99projects 1001 users
127.722s, 145.680s, 131.808s

Optimized
99 projects 1001 users
5.662s, 6.492s, 5.956s, 5.889s, 5.792s

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -30,7 +30,7 @@
         <%= label_tag :project, "Project", class: "col-lg-2 control-label" %>
         <% projects = [['', 0]] + Project.pluck(:name, :id) +
-           Project.deleted.map { |p| ["(deleted) #{p[0]}", p[1]] } %>
+           Project.deleted.pluck(:name, :id).map { |p| ["(deleted) #{p[0]}", p[1]] } %>
```
FYI can do `.map { |name, id| }` for readability

grosser commented on app/models/user.rb:
```diff
@@ -54,25 +54,62 @@ def administrated_projects
+      total = users.count + UserProjectRole.joins(:user, :project).count
+    end
+    summary = ["-", "Generated At", datetime, "Users", users.count.to_s, "Total entries", total.to_s]
+    options_applied = [
```
nice to see this in the dump for later reproduction :)

grosser commented on app/views/csv_exports/index.html.erb:
```diff
@@ -1,8 +1,9 @@
-  <div class="deploy-csv">
-    <%= link_to "Start a CSV export job", new_csv_export_path %>
+  <div>
+    <p><%= link_to "Create Deploys Report", new_csv_export_path %></p>
```
could remove the `Create` part since hitting that link does not create it ... `New Deploys Report` or just `Deploys Report`

grosser commented on app/models/user.rb:
```diff
@@ -54,25 +54,62 @@ def administrated_projects
 
-  def self.to_csv
-    @users = User.order(:id)
+  def self.to_csv(
```
this whole section might make a nice `UserCsvPresenter` class or something like that ... but can refactor that after it's merged ...

rikeoka replied:
should the csv_line and effective_project_role also be moved to the presenter given they aren't using the native model methods.  It'll keep the optimized logic outside of the model, that won't matter for other use cases.

grosser replied:
sounds good!

On Tue, Aug 2, 2016 at 1:51 PM, Robert Ikeoka <notifications@github.com>
wrote:

> In app/models/user.rb
> <https://github.com/zendesk/samson/pull/1132#discussion_r73234742>:
>
> > @@ -54,25 +54,62 @@ def administrated_projects
> >      scope
> >    end
> >
> > -  def self.to_csv
> > -    @users = User.order(:id)
> > +  def self.to_csv(
>
> should the csv_line and effective_project_role also be moved to the
> presenter given they aren't using the native model methods. It'll keep the
> optimized logic outside of the model, that won't matter for other use cases.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1132/files/13cf06f9fb82ef8dddf922856ec64e87df5cea81#r73234742>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZw0o-olHVBP6RncTjlcXq8dq2WOpks5qb622gaJpZM4JK04R>
> .
>

grosser commented on app/views/csv_exports/new.html.erb:
```diff
@@ -43,11 +42,9 @@
+  <div class="clearfix">
+    <hr />
+    <div><strong>Other reports</strong></div>
+    <div><%= link_to "Create Users Report", new_csv_export_path(type: :users) %></div>
```
same here: `Request Users Report` or `Users Report` or `New Users Report`

grosser commented on app/views/csv_exports/new_users.html.erb:
```diff
@@ -0,0 +1,47 @@
+  <div class="clearfix">
+    <hr />
+    <div><strong>Other reports</strong></div>
+    <div><%= link_to "Create Deploys Report", new_csv_export_path %></div>
```
same with the create

grosser commented:
This N+1 optimization work might get lost when someone refactors the code ... so adding a `assert_sql_queries` test could make sure it does not degrade.

👍 looks scary, but reasonable ...

rikeoka commented:
Seeing if this turns green.  
Refactored all of the User Csv methods into the UserCsvPresenter class in /app/presenters/.  Not sure if we'll use API presenters if so maybe it should be in /app/presenters/csv/?  

Added a test for accuracy on the entire to_csv method for inherited option to see if it matches the slow algorithm.

rikeoka commented:
All tests green now.

grosser commented on app/presenters/user_csv_presenter.rb:
```diff
@@ -0,0 +1,99 @@
+  ## - project: Project object or nil for System level role
+  ## - project_role_id: prefetched user_project_role for project
+  ## - project_role_id is prefetched inside of to_csv to optimize to O(N) from O(N*M) using native
+  ##   ActiveRecord methods
```
❤️ great docs!

grosser commented:
👍 looks great :)
